---
title: An R Markdown document converted from "using_singleR_to_perform_annotations.ipynb"
output: html_document
---

# Using SingleR to Perform Annotations:

  - SingleR is a computational method for unbiased cell type recognition of scRNA-seq.
  - SingleR leverages reference transcriptomic datasets of pure cell types to infer the cell of origin of each of the single cells independently.
  - Typically used to compare the gene expression profile of individual cells to a reference database of known cell types in order to assign cell type labels to the cells in the scRNA-seq dataset.
  - Reference object in SingleR is usually a bulk transcriptome dataset, representing the gene expression profiles of different cell types collected from bulk RNA sequencing of pure cell populations.
  - Reference dataset contains the average gene expression values for each gene across the cells of each known cell type.
  - SingleR then uses these average expression profiles to match the gene expression of individual cells from the scRNA-seq dataset to the reference profiles and infer their cell types.

## Loading the Necessary Packages:

```{r}
setwd("/home/maryam/Desktop/Multiome_GEX_data/using_singleR_to_perform_annotation/")

options(repr.plot.width=12, repr.plot.height=6)

options(scipen=100) #avoid scientific notation of numbers
```

```{r}
library(SeuratDisk)
library(Seurat)
library(matrixStats)
library(ggplot2)
library(pheatmap)
library(reshape2)
library(dplyr)
library(tidyr)
library(viridis)
library(RColorBrewer)

#Loading in Some extra packages that are necessary:
library(SingleR)
library(scater)
```

## Load in the GEX data and Marieke's Reference Data:

```{r}
#Load in the GEX object:
seurobj_gex <- LoadH5Seurat("my_annotated_object.h5seurat")
```

```{r}
#Load in Marieke's Reference data:
seurobj_reference <- readRDS("lav_tuned_12-12-2022.rds")
```

```{r}
seurobj_gex@reductions
```

```{r}
seurobj_reference@reductions
```

```{r}
seurobj_gex
```

```{r}
seurobj_reference
```

```{r}
DefaultAssay(seurobj_reference) <- 'RNA'
```

```{r}
seurobj_reference
```

## Try on Pseudobulk Reference Data:

  - First try and make a pseudobulk dataset from the single cell reference data.
  - This means that gene expression will be averaged per cluster.
  - We can try averaging using the *AverageExpression function from the Seurat package.

```{r}
#First set the Idents:
Idents(seurobj_reference) <- seurobj_reference$annotation
pseudobulk_reference <- AverageExpression(seurobj_reference, assays = "RNA", return.seurat = TRUE, verbose = TRUE)
```

```{r}
#Check if there are any 'unknowns':
levels(pseudobulk_reference)
```

  - All the cell types I'm interested in are present.
  - Next step is to convert both dataset to an *sce* object.

```{r}
sce_gex <- as.SingleCellExperiment(seurobj_gex)
sce_pseudobulk <- as.SingleCellExperiment(pseudobulk_reference)
```

```{r}
sce_gex
```

```{r}
assay(sce_gex)
```

```{r}
assay(sce_pseudobulk)
```

```{r}
logcounts(sce_gex) <- as(logcounts(sce_gex), "dgCMatrix")
```

```{r}
assay(sce_gex)
```

```{r}
logcounts(sce_pseudobulk) <- as(logcounts(sce_pseudobulk), "dgCMatrix")
```

## Perform The Annotation:

```{r}
ptm <- proc.time()

intgr_pseudo <- SingleR(test = sce_gex, ref = sce_pseudobulk, labels = colnames(sce_pseudobulk))

proc.time() - ptm
```

## Check the label transfer of the Pseudobulk Integration:

```{r}
colnames(intgr_pseudo)
```

```{r}
#Check the initial colnames of the pseudobulk reference:
colnames(pseudobulk_reference)
```

```{r}
#View the new predicted labels:
table(intgr_pseudo$labels)
```

### Some Insights:

  - Interestingly, unlike the previous methods I'd used (transfer labels/ Mapquery, etc), using SingleR has resulted in the retainment of all 21 different celltypes!
  - Compare to the output recorded in the j. notebook:
      - http://localhost:8888/notebooks/Desktop/marieke_data/lab%20Meeting%20Update%20-%209-06-23.ipynb

```{r}
#Plot findings:
options(repr.plot.width=15, repr.plot.height=10)

plotScoreHeatmap(intgr_pseudo, order.by = "labels", show.labels = F)
```

```{r}
## Remove low quality predictions, i.e 'pruned.labels'
to.remove <- is.na(intgr_pseudo$pruned.labels)
table(Label=intgr_pseudo$labels, Removed=to.remove)
```

```{r}
options(repr.plot.width=12, repr.plot.height=12)

plotDeltaDistribution(intgr_pseudo, ncol=4, dots.on.top = F, show = "delta.med")
```

#### I need to use a gex-object with a umap

```{r}
intgr_pseudo
```

  - Since the output is in the form of a dataframe, integrate it back into my initial gex seurat object.

```{r}
labels <- intgr_pseudo[,4, drop=F]
```

```{r}
seurobj_gex$singleR_reference_bulk <- labels[match(colnames(seurobj_gex), rownames(labels)),1]
```

```{r}
seurobj_gex$singleR_reference_bulk <- factor(seurobj_gex$singleR_reference_bulk, levels = c('ETP','Committed','Specified','TSPp','B_Selection','LMPP','Immature_B_Cells','TSP','pDC','GMP_IRF8hi','GMP_IRF8lo','Erythroid','preDC2','Mature_B_Cells','preDC1','Monocytes','DC2','DC1','Plasma_Cells','NK','Mature_T_Cells'))
```

```{r}
levels(seurobj_gex$singleR_reference_bulk)
```

```{r}
seurobj_gex
```

```{r}
head(seurobj_gex@meta.data)
```

```{r}
### Save this Seurat Object:
SaveH5Seurat(seurobj_gex, "singleR_annotated_motsp_object.h5seurat", overwrite = TRUE)
```

```{r}
#Visualize the plots:

options(repr.plot.width=15, repr.plot.height=10)

a <- DimPlot(seurobj_gex, group.by="predicted.id", reduction="umap_CC", shuffle=T)+labs(title='UMAP based on TransferLabels Output')
b <- DimPlot(seurobj_gex, group.by="singleR_reference_bulk", reduction="umap_CC", shuffle=T)+labs(title='UMAP based on SingleR Output')
a + b 
```

```{r}
b <- DimPlot(seurobj_gex, group.by="singleR_reference_bulk", reduction="umap_CC", shuffle=T, label = FALSE)+labs(title='UMAP based on SingleR Output')
a+b 
```

```{r}
b #use this one
```

```{r}
b
```

